### DISCUSSION TITLE
Nice error message when LSP not found in PATH

### DISCUSSION THREAD

**Discussion by unhammer** [joaotavora/eglot#1318] [Ideas]
I have `  (add-hook 'js-ts-mode-hook #'eglot-ensure)` but I don't have `typescript-language-server` in my global PATH, so whenever I open a js file I see a Message like 

```
Error in post-command-hook (#[0 "\303\301!\205\0r\301q\210\304\305\300\242\306#\210
?\205\0\307\310\311 \")\207" [(#0) #<buffer foo.js> eglot--managed-mode buffer-live-p remove-hook post-command-hook t apply eglot--connect eglot--guess-contact] 4]): (file-missing "Searching for program" "No such file or directory" "typescript-language-server")
``` 

which is kind of noisy and hard to read. It'd be nicer if eglot-ensure just said `Language server typescript-language-server not found in PATH, eglot will be turned off` or something friendler than that. In particular because this might be the first message someone sees after installing eglot for the first time and not having everything fully set up :-)

**Comment by joaotavora** [comment]
It would be nicer, but it would also be nicer if so many people using `eglot-ensure` would stop using it and just use `M-x eglot`.

So many Reddit recipes being copied around and `eglot-ensure` is specifically described like this in the manual:
```
The function eglot-ensure will start an Eglot session for each buffer in which foo-mode is turned on, if there isn’t already an Eglot session that handles the buffer. Note that this variant of starting an Eglot session is non-interactive, so it should be used only when you are confident that Eglot can be started reliably for any file which may be visited with the major-mode in question.
```

So if you are _not confident_ of that, don't use it.

But I'll see what I can do.  In the meantime, you can always define a `my-eglot-ensure` function wrapper which catches errors and demotes them to warnings.

>  In particular because this might be the first message someone sees after installing eglot for the first time and not having everything fully set up :-)

If they are brainless reddit copy-all-the-things newbies, I guess yes.

EDIT: Reddit users are not necessarily brainless, so I hereby apologize.  But they use a platform where popular upvoted things, not scientifically tested things, bubble up and become truth.  Great for spreading misleading ideas, fake news, etc. I happen not to like that, in software and so many other fields.  I prefer the pre-post-truth world where our brains are more actively engaged in healthy scrutiny.

**Reply by unhammer** [reply]
There is a perfectly fine reason that people recommend `eglot-ensure` – it means that you don't have to manually `M-x eglot` for every single buffer of a project. I can't put `eglot` in a mode hook. I would like to have LSP support automatically turned on in projects where there is a language server set up, but at the same time not have to deal with warnings when I just open some random file where LSP isn't set up. I *want* it to be non-interactive.

(Btw, if you really want to recommend users against `eglot-ensure`, maybe warn about it in the docstring? Currently it just says 
`"Start Eglot session for current buffer if there isn't one."`)

**Reply by joaotavora** [reply]
> There is a perfectly fine reason that people recommend eglot-ensure – it means that you don't have to manually M-x eglot for every single buffer of a project

Yes, but no.  Because that is not true.  Just try it.

**Reply by unhammer** [reply]
Aha, I learn something new, thank you. So one `M-x eglot` per project is enough to get it going within that project; perhaps I got confused by trying it with files of different projects.

**Reply by joaotavora** [reply]
Good to have clarified that.  I'll try that idea of demoting the error to warning so it shows up in the `*Warnings*` buffer, and follow your suggestion of more forcefully clarifying that `eglot-ensure` might not be a good idea (a common problem is trying to start Eglot up in an humongous or poorly setup project by accident and then complaining that "Eglot locks up my Emacs")

**Comment by joaotavora** [comment]
I changed `eglot-ensure` to demote errors to warnings. See https://git.savannah.gnu.org/cgit/emacs.git/commit/?id=3ab99e977db01939cc42f285d5ce58807cf2e7d0

In that commit I also expanded the manual even more to clarify that one does NOT need to invoke `M-x eglot` for every buffer in a project.

**Reply by unhammer** [reply]
That was quick :) Just tried it, that looks much nicer: `[eglot] (warning) Searching for program: No such file or directory, typescript-language-server` :-)

**Reply by fuzunspm** [reply]
> That was quick :) Just tried it, that looks much nicer: `[eglot] (warning) Searching for program: No such file or directory, typescript-language-server` :-)

I think you should use exec-path-from-shell-initialize
